File: Exercise_2.cpp
// Time Complexity : O(1)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode : Couldn't find a leetcode problem
// Any problem you faced while coding this : no

#include <bits/stdc++.h>
using namespace std;

// A structure to represent a stack
class StackNode
{
public:
    int data;
    StackNode *next;
};

StackNode *newNode(int data)
{
    StackNode *stackNode = new StackNode();
    stackNode->data = data;
    stackNode->next = NULL;
    return stackNode;
}

int isEmpty(StackNode *root)
{
    // Your code here
    return !root;
}

void push(StackNode **root, int data)
{
    // Your code here
    StackNode *node = newNode(data);
    node->next = *root;
    *root = node;
    cout << data << " pushed to stack\n";
}

int pop(StackNode **root)
{
    // Your code here
    if (isEmpty(*root))
    {
        cout << "Stack is empty\n";
        return -1;
    }
    StackNode *node = *root;
    *root = (*root)->next;
    int pop_val = node->data;
    delete node;
    return pop_val;
}

int peek(StackNode *root)
{
    // Your code here
    if (isEmpty(root))
    {
        cout << "Stack is empty\n";
        return -1;
    }
    return root->data;
}

int main()
{
    StackNode *root = NULL;

    push(&root, 10);
    push(&root, 20);
    push(&root, 30);

    cout << pop(&root) << " popped from stack\n";

    cout << "Top element is " << peek(root) << endl;

    return 0;
}

File: Exercise_2.py

class Node:
    def __init__(self, data):
       self.data = data
       self.next = None
 
class Stack:
    def __init__(self):
        self.root = None
        
    def push(self, data):
        new_node = Node(data)
        new_node.next = self.root
        self.root = new_node
        print(f"{data} pushed to stack")
        
    def pop(self):
        if self.root is None:
            return None
        popped = self.root.data
        self.root = self.root.next
        return popped
        
a_stack = Stack()
while True:
    #Give input as string if getting an EOF error. Give input like "push 10" or "pop"
    print('push <value>')
    print('pop')
    print('quit')
    do = input('What would you like to do? ').split()
    #Give input as string if getting an EOF error. Give input like "push 10" or "pop"
    operation = do[0].strip().lower()
    if operation == 'push':
        a_stack.push(int(do[1]))
    elif operation == 'pop':
        popped = a_stack.pop()
        if popped is None:
            print('Stack is empty.')
        else:
            print('Popped value: ', int(popped))
    elif operation == 'quit':
        break
